File: IO.Astrodynamics/Maneuvers/CombinedManeuver.cpp
/*
 Copyright (c) 2021-2023. Sylvain Guillet (sylvain.guillet@tutamail.com)
 */
#include <CombinedManeuver.h>
#include <ConicOrbitalElements.h>
#include <Vector3D.h>
#include <Parameters.h>

#include <utility>

IO::Astrodynamics::Maneuvers::CombinedManeuver::CombinedManeuver(std::vector<IO::Astrodynamics::Body::Spacecraft::Engine*> engines, IO::Astrodynamics::Propagators::Propagator &propagator, const double inclination, const double perigeeRadius) : IO::Astrodynamics::Maneuvers::ManeuverBase(std::move(engines), propagator), m_inclination{inclination}, m_peregeeRadius{perigeeRadius}
{
}

IO::Astrodynamics::Maneuvers::CombinedManeuver::CombinedManeuver(std::vector<IO::Astrodynamics::Body::Spacecraft::Engine*> engines, IO::Astrodynamics::Propagators::Propagator &propagator, const double inclination, const double perigeeRadius, const IO::Astrodynamics::Time::TDB &minimumEpoch) : IO::Astrodynamics::Maneuvers::ManeuverBase(std::move(engines), propagator, minimumEpoch), m_inclination{inclination}, m_peregeeRadius{perigeeRadius}
{
}

void IO::Astrodynamics::Maneuvers::CombinedManeuver::Compute(const IO::Astrodynamics::OrbitalParameters::OrbitalParameters &orbitalParams)
{
    //Compute delta V vector
    m_deltaV = std::make_unique<IO::Astrodynamics::Math::Vector3D>(GetDeltaV(orbitalParams.ToStateVector()));
}

IO::Astrodynamics::OrbitalParameters::StateOrientation IO::Astrodynamics::Maneuvers::CombinedManeuver::ComputeOrientation(const IO::Astrodynamics::OrbitalParameters::OrbitalParameters &maneuverPoint)
{
    auto maneuverVelocity = GetDeltaV(maneuverPoint.ToStateVector());

    return IO::Astrodynamics::OrbitalParameters::StateOrientation{maneuverVelocity.Normalize().To(m_spacecraft.Front), IO::Astrodynamics::Math::Vector3D(0.0, 0.0, 0.0), maneuverPoint.GetEpoch(), maneuverPoint.GetFrame()};
}

IO::Astrodynamics::Math::Vector3D IO::Astrodynamics::Maneuvers::CombinedManeuver::GetDeltaV(const IO::Astrodynamics::OrbitalParameters::StateVector &sv) const
{
    double e{};
    double rp{};
    double meanAnomaly = sv.GetMeanAnomaly();
    double periapsisArgument = sv.GetPeriapsisArgument();
    double apogee=sv.GetApogeeVector().Magnitude();

    //If target perigee is higher than current apogee
    if (m_peregeeRadius > apogee)
    {
        rp = apogee;
        e = 1 - (2 / ((m_peregeeRadius / rp) + 1));

        //Periapse argument will turn by 180°
        meanAnomaly = std::fmod(meanAnomaly+=Constants::PI,Constants::_2PI);
        periapsisArgument += IO::Astrodynamics::Constants::PI;
    }
    else
    {
        rp = m_peregeeRadius;
        e = 1 - (2 / ((apogee / rp) + 1));
    }

    auto targetOrbit = IO::Astrodynamics::OrbitalParameters::ConicOrbitalElements(sv.GetCenterOfMotion(), rp, e, m_inclination, sv.GetRightAscendingNodeLongitude(), periapsisArgument, meanAnomaly, sv.GetEpoch(), sv.GetFrame());

    return targetOrbit.ToStateVector().GetVelocity() - sv.GetVelocity();
}

IO::Astrodynamics::Math::Vector3D
IO::Astrodynamics::Maneuvers::CombinedManeuver::ManeuverPointComputation(const IO::Astrodynamics::OrbitalParameters::OrbitalParameters &orbitalParameters)
{
    return orbitalParameters.GetApogeeVector();
}

File: IO.Astrodynamics.Net/IO.Astrodynamics/Maneuver/CombinedManeuver.cs
using System;
using IO.Astrodynamics.Body;
using IO.Astrodynamics.Body.Spacecraft;
using IO.Astrodynamics.Math;
using IO.Astrodynamics.OrbitalParameters;
using IO.Astrodynamics.Propagator.Events;
using IO.Astrodynamics.TimeSystem;

namespace IO.Astrodynamics.Maneuver
{
    public class CombinedManeuver : ImpulseManeuver
    {
        public double TargetPerigeeHeight { get; } = double.NaN;
        public double TargetInclination { get; } = double.NaN;

        public CombinedManeuver(CelestialItem maneuverCenter, Time minimumEpoch, TimeSpan maneuverHoldDuration, double perigeeRadius, double inclination, Engine engine) : base(
            maneuverCenter, minimumEpoch, maneuverHoldDuration, engine)
        {
            TargetPerigeeHeight = perigeeRadius;
            TargetInclination = inclination;
        }

        /// <summary>
        /// Fires at apogee: r·v = 0, transitioning from positive to negative.
        /// </summary>
        public override double ComputeEventValue(StateVector localState) => localState.Position * localState.Velocity;
        public override CrossingDirection EventCrossingDirection => CrossingDirection.PositiveToNegative;

        /// <summary>
        /// Apogee must be nearly aligned with ascending node (dot product > 0.9).
        /// </summary>
        public override bool CheckPreconditions(StateVector state) =>
            System.Math.Abs(state.ApogeeVector().Normalize() * state.AscendingNodeVector().Normalize()) > 0.9;

        protected override Vector3 Execute(StateVector stateVector)
        {
            double e;
            double meanAnomaly = stateVector.MeanAnomaly();
            double periapsisArgument = stateVector.ArgumentOfPeriapsis();
            double apogee = stateVector.ApogeeVector().Magnitude();

            //If target perigee is higher than current apogee
            if (TargetPerigeeHeight > apogee)
            {
                e = 1.0 - (2.0 / ((TargetPerigeeHeight / apogee) + 1.0));

                //Periapse argument will turn by 180°
                meanAnomaly = (meanAnomaly + Constants.PI) % Constants._2PI;
                periapsisArgument += Constants.PI;
            }
            else
            {
                e = 1.0 - (2.0 / ((apogee / TargetPerigeeHeight) + 1.0));
            }

            var targetOrbit = new KeplerianElements((TargetPerigeeHeight + apogee) * 0.5, e, TargetInclination, stateVector.AscendingNode(), periapsisArgument, meanAnomaly,
                stateVector.Observer, stateVector.Epoch, stateVector.Frame, perigeeRadius: stateVector.PerigeeRadius());

            return targetOrbit.ToStateVector().Velocity - stateVector.Velocity;
        }
    }
}